pytensor/graph/basic.py

```diff
@@ -970,6 +970,136 @@ def applys_between(
     )
 
 
+def truncated_graph_inputs(
+    outputs: Sequence[Variable], conditions: Optional[Collection[Variable]] = None
+) -> List[Variable]:
+    """Get the conditional subset for outputs.
+
+
+    Conditional subset is a minimal set of variables required to compute outputs.
+
+    Parameters
+    ----------
+    outputs : Collection[Variable]
+        Variable to get conditions for
+    conditions : Optional[Collection[Variable]]
+        Additional conditions to assume, by default None
+
+    Returns
+    -------
+    List[Variable]
+        Variables required to compute ``outputs``
+
+    Examples
+    --------
+    The returned nodes marked in (parenthesis), condition nodes are ``c``, output nodes are ``o``
+
+    * No conditions
+
+    .. code-block::
+
+        n - n - (o)
+
+    * One condition
+
+    .. code-block::
+
+        n - (c) - o
+
+    * Two conditions where on depends on another, both returned
+
+    .. code-block::
+
+        (c) - (c) - o
+
+    * Additional nodes are present
+
+    .. code-block::
+
+           (c) - n - o
+        n - (n) -'
+
+    * Disconnected condition not returned
+
+    .. code-block::
+
+        (c) - n - o
+         c
+
+    * Disconnected output is present and returned
+
+    .. code-block::
+
+        (c) - (c) - o
+        (o)
+
+    * Condition on itself adds itself
+
+    .. code-block::
+
+        n - (c) - (o/c)
+
+    """
+    # simple case, no additional conditions
+    blockers: Set[Variable] = set()
+    independent_nodes = list()
+    # blockers have known independent nodes and pre-conditions
+    candidates = list(outputs)
+    if not conditions:  # None or empty
+        # just filter out unique variables
+        for node in candidates:
+            if node not in blockers:
+                independent_nodes.append(node)
+                blockers.add(node)
+        # no more actions are needed
+        return independent_nodes
```

Comment:
Shouldn't it be the inputs of the outputs? If you call `replace(x, {x: new_x})`, I imagine it will fail because this will return `x`?

Reply:
I just added a test to show it will also work as expected